### ISSUE TITLE
Matching issue introduced in last version of PCRE2 library ( {x} doesn't match, {X-1,X+1} does )

### ISSUE THREAD

**Issue by peter-thoenen** [PhilipHazel/pcre2#82]
Came up with another project which led to it being an issue with the library itself as it also fails with pcre2test.  For a full discussion see this bug:

https://github.com/firasdib/Regex101/issues/1704#issuecomment-1019503942
https://github.com/php/php-src/issues/7994

Basically given "aa" [a]{2} fails, [a]{1,3} works.

Sorry took a weird path to end up at this place as didn't realized this was just a library everybody used lol.

**Comment by carenas** [comment]
issue released with 10.34 and introduced in commit bf15267c30e202e3192472adebc8865915a0c72b
Author: Philip.Hazel <Philip.Hazel@gmail.com>
Date:   Mon Sep 9 17:00:19 2019 +0000

    Optimize classes such as [Aa] to be a single caseless character.

 ChangeLog                |  6 +++-
 src/pcre2_compile.c      | 85 ++++++++++++++++++++++++++++++++++++++----------
 testdata/testinput10     |  2 ++
 testdata/testinput12     |  2 ++
 testdata/testinput2      |  2 ++
 testdata/testoutput10    |  7 ++++
 testdata/testoutput12-16 |  7 ++++
 testdata/testoutput12-32 |  7 ++++
 testdata/testoutput2     | 14 ++++++++
 9 files changed, 114 insertions(+), 18 deletions(-)

and only affects the matching when the first character in the class doesn't correspond to the correct case that is shown in the data, but ironically it is because the last character in the class is registered as literal (and not caseless)

```
printf "/[Aa]{2}/\naa\n" | pcre2test
PCRE2 version 10.39 2021-10-29
/[Aa]{2}/
aa
No match
```

**Comment by PhilipHazel** [comment]
This was an optimization bug. [Aa] is turned into caseless 'A' but when it was last in a pattern, the "must have this character" optimization was not getting flagged as caseless. I have committed a small patch that fixes this. Thanks for the report and triage.

**Comment by guilliamxavier** [comment]
For the record: the fix is commit https://github.com/PCRE2Project/pcre2/commit/fdd94791086e466835ed61640cf31380efd1fbc5 (released in version 10.40)
